File: +bsv/plotInjectionsCombined.m
function plotInjectionsCombined(experimentImgs, allenAtlasPath, inputRegions, numberOfSlices, numberOfPixels, plane, regionOnly, smoothing, colorLimits, color, normalizationMethod, experimentRegionInfo, atlasType, atlasResolution)
% plotInjectionsCombined - Plot injection sites for multiple regions in ONE figure
%
% Creates a single figure with one row per input region showing injection sites
% This function builds the plot manually without calling plotConnectivity

% Handle optional atlas parameters
if nargin < 13 || isempty(atlasType)
    atlasType = 'allen'; % Default to Allen atlas
end
if nargin < 14 || isempty(atlasResolution)
    atlasResolution = 10; % Default to 10um resolution
end

%% Construct atlas filenames based on type and resolution
switch lower(atlasType)
    case 'allen'
        if atlasResolution == 10
            annotationFile = 'annotation_volume_10um_by_index.npy';
            structureFile = 'structure_tree_safe_2017.csv';
        elseif atlasResolution == 20
            annotationFile = 'annotation_volume_v2_20um_by_index.npy';
            structureFile = 'UnifiedAtlas_Label_ontology_v2.csv';
        else
            error('Unsupported Allen atlas resolution: %d. Supported resolutions are 10 and 20 um.', atlasResolution);
        end
    otherwise
        % For custom atlases, construct filename from type and resolution
        annotationFile = sprintf('%s_annotation_%dum.npy', atlasType, atlasResolution);
        structureFile = sprintf('%s_structure_tree.csv', atlasType);
end

%% Load atlas
av = readNPY([allenAtlasPath, filesep, annotationFile]);
st = loadStructureTree([allenAtlasPath, filesep, structureFile]);
atlas_slice_spacing = atlasResolution; % Use the provided resolution

%% Create single combined figure
mainFig = figure('Name', 'Combined Injection Sites - All Regions', 'Color', 'w');
nRegions = length(inputRegions);

fprintf('Creating ONE combined figure for %d regions...\n', nRegions);

%% Process each region and create subplots in the main figure
for iRegion = 1:nRegions
    fprintf('Adding region %d/%d: %s to combined plot\n', iRegion, nRegions, inputRegions{iRegion});
    
    % Find structure indices for this region
    curr_plot_structure_idx = find(contains(st.acronym, inputRegions{iRegion}));
    keepStruct = false(size(curr_plot_structure_idx, 1), 1);
    for iCurrStruct = 1:size(curr_plot_structure_idx, 1)
        curr_struct = st.acronym{curr_plot_structure_idx(iCurrStruct)};
        keepStruct(iCurrStruct) = strcmp(curr_struct(1:length(inputRegions{iRegion})), inputRegions{iRegion});
    end
    curr_plot_structure_idx = curr_plot_structure_idx(keepStruct);
    
    if isempty(curr_plot_structure_idx)
        fprintf('Warning: No structure found for region %s\n', inputRegions{iRegion});
        continue;
    end
    
    plot_structure_color = hex2dec(reshape(st.color_hex_triplet{curr_plot_structure_idx(1)}, 2, [])') ./ 255;
    
    % Get chunk limits for this region
    structureLimits = find(ismember(av(:, :, 1:1:end/2), curr_plot_structure_idx));
    [APvalues, ~, MLvalues] = ind2sub(size(av), structureLimits);
    
    if strcmp(plane, 'coronal')
        curr_limits = [min(APvalues), max(APvalues)];
        chunks_region = curr_limits(1):(curr_limits(2) - curr_limits(1)) / numberOfSlices:curr_limits(2);
    else
        curr_limits = [min(MLvalues), max(MLvalues)];
        chunks_region = curr_limits(1):(curr_limits(2) - curr_limits(1)) / numberOfSlices:curr_limits(2);
    end
    
    % Collapse hemispheres for data
    halfSlices = floor(size(experimentImgs, 3)/2);
    theseLocations_ap_dv_ml = experimentImgs(:, :, 1:halfSlices) + experimentImgs(:, :, end:-1:end-halfSlices+1);
    
    % Process each chunk (slice) for this region
    for iChunk = 1:numberOfSlices
        % Calculate subplot position: nRegions rows, numberOfSlices columns
        subplotPos = (iRegion - 1) * numberOfSlices + iChunk;
        subplot(nRegions, numberOfSlices, subplotPos);
        hold on;
        
        % Get structure boundaries for this chunk
        if strcmp(plane, 'coronal')
            region_area = permute(ismember(av(round(chunks_region(iChunk)):1:round(chunks_region(iChunk+1)), ...
                1:1:end, 1:1:end/2), curr_plot_structure_idx), [3, 1, 2]);
        else
            region_area = permute(ismember(av(1:1:end, ...
                1:1:end, round(chunks_region(iChunk)):1:round(chunks_region(iChunk+1))), ...
                curr_plot_structure_idx), [3, 1, 2]);
        end
        
        [regionLocation(1, :), regionLocation(2, :), regionLocation(3, :)] = ind2sub(size(region_area), find(region_area));
        
        if strcmp(plane, 'coronal')
            regionLocation(2, :) = regionLocation(2, :) + round(chunks_region(iChunk)) - 1;
        else
            regionLocation(1, :) = regionLocation(1, :) + round(chunks_region(iChunk)) - 1;
        end
        
        % Get projection data for this chunk
        thisdiff = mean(diff(chunks_region(:)));
        
        if strcmp(plane, 'coronal')
            % Create bins for this chunk
            x_range = [min(regionLocation(1, :)), max(regionLocation(1, :))];
            y_range = [min(regionLocation(3, :)), max(regionLocation(3, :))];
            
            if diff(x_range) > 0 && diff(y_range) > 0
                xEdges = linspace(x_range(1), x_range(2), numberOfPixels+1);
                yEdges = linspace(y_range(1), y_range(2), numberOfPixels+1);
                
                % Extract and average data
                dataSlice = theseLocations_ap_dv_ml(round((chunks_region(iChunk) - thisdiff)/10):round((chunks_region(iChunk) + thisdiff)/10), ...
                    round(yEdges/10), round(xEdges/10));
                meanData = nanmean(dataSlice, 1);
                projtemp = permute(squeeze(meanData), [2, 1]);
                
                % Create mask for region boundary
                [X, Y] = meshgrid(xEdges(1:end-1), yEdges(1:end-1));
                if size(regionLocation, 2) > 2
                    boundary_idx = boundary(regionLocation(1, :)', regionLocation(3, :)', 0);
                    mask = inpolygon(X, Y, regionLocation(1, boundary_idx), regionLocation(3, boundary_idx));
                    projtemp(~mask') = NaN;
                end
                
                % Plot the data
                imagesc(xEdges(1:end-1), yEdges(1:end-1), projtemp');
                
                % Plot region boundary
                if size(regionLocation, 2) > 2
                    plot(regionLocation(1, boundary_idx), regionLocation(3, boundary_idx), ...
                        'Color', plot_structure_color, 'LineWidth', 2);
                end
                
                % Set colormap and limits
                colormap(gca, flipud(gray(256)));
                set(gca, 'color', [0.5, 0.5, 0.5]);
                clim([0, 1]);
                
                % Make NaN values grey
                set(gca, 'color', [0.5, 0.5, 0.5]);
            end
        end
        
        % Format axes
        axis equal; axis tight; axis off;
        if strcmp(plane, 'coronal')
            set(gca, 'YDir', 'reverse');
        end
        
        % Add labels
        if iChunk == 1
            ylabel(sprintf('%s', inputRegions{iRegion}), 'FontWeight', 'bold', 'FontSize', 10, 'Rotation', 90);
        end
        
        if iRegion == 1
            slice_ARA = round(mean(chunks_region(iChunk:iChunk+1))/10);
            title(sprintf('ARA %d', slice_ARA), 'FontSize', 9);
        end
    end
end

% Add overall title and colorbar
sgtitle('Injection Sites by Region', 'FontSize', 14, 'FontWeight', 'bold');

% Add colorbar
subplot(nRegions, numberOfSlices, nRegions * numberOfSlices);
cb = colorbar('eastoutside');
cb.Label.String = 'Injection Intensity';
cb.Label.FontSize = 9;
cb.Label.Rotation = 270;
cb.Label.VerticalAlignment = 'bottom';

fprintf('Combined injection plot completed in ONE figure!\n');

end

File: bsv/plot_injections_combined.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.spatial import ConvexHull
from matplotlib.path import Path

from .atlas_utils import load_atlas, find_structure_indices, get_structure_color


def plot_injections_combined(experiment_imgs, allen_atlas_path, input_regions,
                              number_of_slices, number_of_pixels, plane,
                              region_only, smoothing, color_limits, color,
                              normalization_method, experiment_region_info=None,
                              atlas_type='allen', atlas_resolution=10):
    av, st = load_atlas(allen_atlas_path, atlas_type, atlas_resolution)
    n_regions = len(input_regions)

    fig, axes = plt.subplots(n_regions, number_of_slices,
                              figsize=(3 * number_of_slices, 3 * n_regions),
                              squeeze=False)
    fig.patch.set_facecolor('white')
    fig.suptitle('Injection Sites by Region', fontsize=14, fontweight='bold')

    # Collapse hemispheres
    half = experiment_imgs.shape[2] // 2
    collapsed = experiment_imgs[:, :, :half] + experiment_imgs[:, :, ::-1][:, :, :half]
    global_vmax = np.nanmax(collapsed) if np.nanmax(collapsed) > 0 else 1

    for i_reg in range(n_regions):
        region = input_regions[i_reg]
        curr_idx = find_structure_indices(st, region)
        if not curr_idx:
            print(f'Warning: No structure found for region {region}')
            continue
        region_color = get_structure_color(st, curr_idx[0])

        # Get structure limits
        half_ml = av.shape[2] // 2
        structure_mask = np.isin(av[:, :, :half_ml], curr_idx)
        ap_vals, _, ml_vals = np.where(structure_mask)

        if plane == 'coronal':
            limits = [ap_vals.min(), ap_vals.max()]
        else:
            limits = [ml_vals.min(), ml_vals.max()]
        step = (limits[1] - limits[0]) / number_of_slices
        chunks = [limits[0] + j * step for j in range(number_of_slices + 1)]

        for i_chunk in range(number_of_slices):
            ax = axes[i_reg, i_chunk]
            chunk_start = int(round(chunks[i_chunk]))
            chunk_end = int(round(chunks[i_chunk + 1]))

            if plane == 'coronal':
                region_area = np.isin(av[chunk_start:chunk_end + 1, :, :half_ml], curr_idx)
                ml_loc, ap_loc, dv_loc = np.where(region_area.transpose(2, 0, 1))
                ap_loc += chunk_start
                bnd_x, bnd_y = ml_loc, dv_loc
            else:
                region_area = np.isin(av[:, :, chunk_start:chunk_end + 1], curr_idx)
                ml_loc, ap_loc, dv_loc = np.where(region_area.transpose(2, 0, 1))
                ml_loc += chunk_start
                bnd_x, bnd_y = ap_loc, dv_loc

            if len(bnd_x) < 3:
                ax.axis('off')
                continue

            x_range = [bnd_x.min(), bnd_x.max()]
            y_range = [bnd_y.min(), bnd_y.max()]
            if x_range[1] - x_range[0] == 0 or y_range[1] - y_range[0] == 0:
                ax.axis('off')
                continue

            x_edges = np.linspace(x_range[0], x_range[1], number_of_pixels + 1)
            y_edges = np.linspace(y_range[0], y_range[1], number_of_pixels + 1)

            this_diff = np.mean(np.diff(chunks))
            if plane == 'coronal':
                ap_s = max(0, int(round((chunks[i_chunk] - this_diff) / 10)))
                ap_e = min(collapsed.shape[0] - 1, int(round((chunks[i_chunk] + this_diff) / 10)))
                y_idx = np.clip((y_edges / 10).astype(int), 0, collapsed.shape[1] - 1)
                x_idx = np.clip((x_edges / 10).astype(int), 0, collapsed.shape[2] - 1)
                data_slice = collapsed[ap_s:ap_e + 1][:, y_idx][:, :, x_idx]
                mean_data = np.nanmean(data_slice, axis=0)
                projtemp = mean_data.T
            else:
                ax.axis('off')
                continue

            # Mask outside region
            try:
                hull = ConvexHull(np.column_stack([bnd_x, bnd_y]))
                hull_pts = np.column_stack([bnd_x, bnd_y])[hull.vertices]
                path = Path(hull_pts)
                mask = np.zeros(projtemp.shape[:2], dtype=bool)
                for ix in range(len(x_edges)):
                    for iy in range(len(y_edges)):
                        mask[ix, iy] = path.contains_point([x_edges[ix], y_edges[iy]])
                projtemp[~mask] = np.nan
            except Exception:
                pass

            ax.imshow(projtemp.T, origin='upper' if plane == 'coronal' else 'lower',
                       extent=[x_edges[0], x_edges[-1], y_edges[-1], y_edges[0]],
                       cmap='gray_r', vmin=0, vmax=global_vmax, aspect='equal')
            ax.set_facecolor('0.5')

            # Plot boundary
            try:
                hull = ConvexHull(np.column_stack([bnd_x, bnd_y]))
                hull_pts_plot = np.column_stack([bnd_x, bnd_y])[np.append(hull.vertices, hull.vertices[0])]
                ax.plot(hull_pts_plot[:, 0], hull_pts_plot[:, 1], color=region_color, linewidth=2)
            except Exception:
                pass

            ax.set_xticks([])
            ax.set_yticks([])
            ax.axis('off')

            if i_chunk == 0:
                ax.set_ylabel(region, fontweight='bold', fontsize=10, rotation=90)
                ax.axis('on')
                ax.set_xticks([])
                ax.set_yticks([])
                for spine in ax.spines.values():
                    spine.set_visible(False)

            if i_reg == 0:
                slice_ara = int(round(np.mean(chunks[i_chunk:i_chunk + 2]) / 10))
                ax.set_title(f'ARA {slice_ara}', fontsize=9)

    plt.tight_layout()
    plt.show(block=False)
